src/react/3-react-theory.md

```diff
@@ -0,0 +1,114 @@
+
+```html title="React has two types of components" subtitle="Class components"
+<div id="root"></div><script crossorigin src="//unpkg.com/react@16/umd/react.development.js"></script><script crossorigin src="//unpkg.com/react-dom@16/umd/react-dom.development.js"></script><script type="jsx">
+
+class MyClassComponent extends React.Component {
+  render() {
+    return <div>hello world</div>;
+  }
+}
+
+ReactDOM.render(<MyClassComponent />,document.getElementById('root'));</script>
+```
+@codepen
+
+### What's the Difference?
+
+No matter which type of component you're working with, they can be rendered the same way:
+
+```jsx title="Both types of components are rendered the same way"
+function App() {
+  return (
+    <div>
+      <MyFuncComponent />
+      <MyClassComponent />
+    </div>
+  );
+}
+```
+
+For all intents and purposes, aside from the syntax there's not much difference between the two.
+
+Originally however, class components were the most popular because they could keep track of state and respond to specific component lifecycle events. This wasn't possible in functional component until the release of the React Hooks API in v16.
+
+Now-a-days both class and functional components are used commonly in code-bases around the world, but there has been more of a shift to using functional components only as the syntax for them is considered more readable and cleaner by many.
+
+For the remainder of this training, we'll be using functional components exclusively!
+
+## How does React work?
+
+React is all about components. In fact, that's the whole point of the library, to help developers create re-usable components easily.
+
+But what actually is a component?
+
+In a nutshell, a component is a modular collection of functionality and state which returns JSX.
+
+- **State** - The state of a component is all of the values or pieces of data that it's keeping track of. The state is generally displayed either directly or indirectly in the JSX. 
+- **Functionality** - The functionality of a component is what it does in response to events. For example a button component might spawn a browser alert whenever it's clicked.
+
+Often times the state and the functionality of a component are combined together. Let's take a simple counter component as an example:
+
+```html
+<div id="root"></div><script crossorigin src="//unpkg.com/react@16/umd/react.development.js"></script><script crossorigin src="//unpkg.com/react-dom@16/umd/react-dom.development.js"></script><script type="jsx">ReactDOM.render(<Counter />,document.getElementById('root'));
+
+function Counter(){
+    // a piece of state called `clickCount` is initialized to 0
+    const [clickCount, setClickCount] = React.useState(0)
+
+    return (
+        <div>
+            <button onClick={() => setClickCount(clickCount + 1)}>+1</button>
+            Clicked {clickCount} times
+        </div>
+    )
+}
+
+</script>
+```
+@highlight 5,9-10,only
```

Comment:
This looks great. I love how you made it look like the powerpoint with the highlighted lines. That adds a lot to the article.